### PR TITLE
Json encode API function

### LINKED ISSUE #5879: Add native function json.encode(table) as a counterpart to json.decode(str)
**Is your feature request related to a problem? Please describe (REQUIRED):**
I would like a way to encode lua tables to JSON in the same way that defold already supports decoding them with json.decode()

**Describe the solution you'd like (REQUIRED):**
Implementation to Defold of a native `json.encode(table) --> str` function that converts lua tables to stringified JSON, in the reverse of how `json.decode(str)` already works

**Describe alternatives you've considered (REQUIRED):**
A manual implementation, and native extension seem like overkill for this feature when the decode counterpart already exists

### PR DESCRIPTION
**Pull request by Jhonnyg**
Added a new function to the json namespace `json.encode(tbl)` that can be used to generate a json formatted string from a lua table:

```
local json_str = json.encode({"test", test_tbl = { 1, 2, 3}})
pprint(json_str) -- prints {"1":"test","test_tbl":[1,2,3]}
```

Fixes #5879

### CODE REVIEW

**Comment by Jhonnyg** [comment]
Q: Right now the cjson source encodes userdata, functions (and threads) as "object at <memory-address>", is that the behaviour we want?
Also encoding a light user data will yield an error, I'm guessing for security reasons but the source doesn't say why

**Comment by britzl** [comment]
> Right now the cjson source encodes userdata, functions (and threads) as "object at ", is that the behaviour we want?

Yeah, maybe. Or nil. Or empty string. But I think "object at" is perfectly acceptable. Maybe an improvement would be if it was "object at" for userdata, "function at" for functions and "thread at" for threads.

> Also encoding a light user data will yield an error, I'm guessing for security reasons but the source doesn't say why

I wonder why it doesn't encode as "object at"?

**Comment by AGulev** [comment]
for the internal engine's functionality it feels like user types from the engine should be serialisable, no? 
maybe we can encode them into json as data_ user_data_type?
for example for vector3:
{x:1,y:1,z:1,_user_type:hash}

**Comment by britzl** [comment]
> for example for vector3:
> {x:1,y:1,z:1,_user_type:hash}

Shouldn't it be `_user_type:vector3` ?

**Comment by AGulev** [comment]
> Shouldn't it be `_user_type:vector3` ?

Whatever we need to be able to serialise <-> deserialise these values

**Comment by Jhonnyg** [comment]
> for the internal engine's functionality it feels like user types from the engine should be serialisable, no?
> maybe we can encode them into json as data_ user_data_type?
> for example for vector3:
> {x:1,y:1,z:1,_user_type:hash}

Hm, they _could_ be serialized, but then we need to check the userdata against all possible type hashes we support in the engine, which are all statically defined in their own modules, i.e script_buffer.cpp:

```
namespace dmScript
{
    static uint32_t SCRIPT_BUFFER_TYPE_HASH = 0;
    static uint32_t SCRIPT_BUFFERSTREAM_TYPE_HASH = 0;
}
```

If we want to support that, we'd need to get those symbols from the respective script modules and I don't know if we want to expose that? Even if we know all the string names of the constructs we supply and could potentially check that in the encode function, I don't think we want to do that in case we do any changes in the type names elsewhere. 

Unless we want to move the hashes per module to something more widely accessible in other parts of the engine, we'd have to expose a function per script module that does the actual check. But can/should we even do serialisation for certain types like a buffer object? Our json.decode doesn't support de-serialisation from these types anyway, right?

File: engine/script/src/luacjson/cjson.c [hunk 1]
```diff
@@ -0,0 +1,1114 @@
...
+#if 0 // Defold: unused
+static json_config_t *cfg;
+static json_config_t *json_fetch_config(lua_State *l)
+{
+    // json_config_t *cfg;
+    cfg = (json_config_t*)lua_touserdata(l, lua_upvalueindex(1));
+    if (!cfg)
+        luaL_error(l, "BUG: Unable to fetch CJSON configuration");
+    return cfg;
+}
```
Comment (Jhonnyg) [inline engine/script/src/luacjson/cjson.c:202]: These functions are all disabled mainly to suppress warnings, the entire file is already heavily modified and will likely be replaced in the future for when we replace our json decoder and move encoding to dmsdk level.

File: engine/script/src/script_json.cpp [hunk 2]
```diff
@@ -24,8 +24,11 @@
 
 extern "C"
 {
-#include <lua/lua.h>
-#include <lua/lauxlib.h>
+    #include <lua/lua.h>
+    #include <lua/lauxlib.h>
+
+    // Defined in luacjson/cjson.c
+    int lua_cjson_encode(lua_State *L);
```
Comment (Jhonnyg) [inline engine/script/src/script_json.cpp:31]: The entry point for the library required us to push both encode and decode functions to a table, so we've added an extra function in the cjson file that exposes the functionality we need instead.

File: engine/script/wscript [hunk 3]
```diff
@@ -58,7 +58,6 @@ def shutdown(self):
 
     serv = None
     if not getattr(waflib.Options.options, 'skip_tests', False):
-        print("SERVER")
```
Comment (Jhonnyg) [inline engine/script/wscript:61]: I think I mistakenly added this when I did the p2->p3 conversion

**Review by JCash — DISMISSED** [review]

File: engine/script/src/luacjson/strbuf.c [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+static void die(const char *fmt, ...)
+{
+    va_list arg;
+
+    va_start(arg, fmt);
+    vfprintf(stderr, fmt, arg);
+    va_end(arg);
+    fprintf(stderr, "\n");
+
+    exit(-1);
```
[the remark this answers is not in the document]
Comment (Jhonnyg) [reply engine/script/src/luacjson/strbuf.c:41]: This isn't our code, but you are right we should probably not exit here at least. As for lua error, I'll see what I can do

**Review by JCash — APPROVED** [review]

**Pull request merged**